File: flutter/packages/hatake_core/lib/src/definition/dashboard_value_definition.dart
import 'package:equatable/equatable.dart';

import 'aggregate_ops.dart';

/// How a `metric` item turns the rows it fetched into one number.
///
/// The framework never queries an aggregate: the [Repository] returns rows and
/// this describes the reduction over them (see `AggregateRegistry`). When the
/// backend already aggregates, use `count`, or point [field] at the column the
/// endpoint returns and fetch a single row.
class DashboardValueDefinition extends Equatable {
  /// Aggregate operation name ([AggregateOps] or a plugin's).
  final String aggregate;

  /// Field to reduce. Not needed by `count`; required by the others.
  final String? field;

  const DashboardValueDefinition({
    this.aggregate = AggregateOps.count,
    this.field,
  });

  @override
  List<Object?> get props => [aggregate, field];
}

File: java/src/main/java/io/hatake/core/DashboardValueDefinition.java
package io.hatake.core;

/**
 * ダッシュボードの {@code metric} カードが、取得した行を1つの数値に畳む方法。
 *
 * <p>集約自体は {@link Aggregates}。Framework は「集計クエリ」を投げない
 * ＝ Repository が行を返し、その行に対する畳み込みだけをここが決める。
 *
 * @param aggregate 集約オペレーション名（{@link Aggregates} の組込み or プラグイン）
 * @param field 畳む項目名。{@code count} では不要、それ以外は必須（無ければ null 結果）
 */
public record DashboardValueDefinition(String aggregate, String field) {
}
